contracts/cw-ics721-bridge/src/ibc.rs

```diff
@@ -0,0 +1,455 @@
+    pub token_ids: Vec<String>,
+    /// URL that points to metadata for each token being
+    /// transfered. `tokenUris[N]` should hold the metadata for
+    /// `tokenIds[N]` and both lists should have the same length.
+    pub token_uris: Vec<String>,
+    /// The address sending the tokens on the sending chain.
+    pub sender: String,
+    /// The address that should receive the tokens on the receiving
+    /// chain.
+    pub receiver: String,
+}
+
+pub const IBC_VERSION: &str = "ics721-1";
+const ACK_ERROR_FALLBACK: &str =
+    "an unexpected error occurred - error text is hidden because it would serialize as ACK success";
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_channel_open(
+    _deps: DepsMut,
+    _env: Env,
+    msg: IbcChannelOpenMsg,
+) -> Result<(), ContractError> {
+    validate_order_and_version(msg.channel(), msg.counterparty_version())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_channel_connect(
+    _deps: DepsMut,
+    _env: Env,
+    msg: IbcChannelConnectMsg,
+) -> Result<IbcBasicResponse, ContractError> {
+    validate_order_and_version(msg.channel(), msg.counterparty_version())?;
+
+    Ok(IbcBasicResponse::new()
+        .add_attribute("method", "ibc_channel_connect")
+        .add_attribute("channel", &msg.channel().endpoint.channel_id)
+        .add_attribute("port", &msg.channel().endpoint.port_id))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_channel_close(
+    _deps: DepsMut,
+    _env: Env,
+    msg: IbcChannelCloseMsg,
+) -> Result<IbcBasicResponse, ContractError> {
+    match msg {
+        IbcChannelCloseMsg::CloseInit { channel: _ } => Err(ContractError::CantCloseChannel {}),
+        IbcChannelCloseMsg::CloseConfirm { channel: _ } => {
+            // TODO: Is this actually the correct logic? If we do hit
+            // this, IBC is telling us "the channel has been closed
+            // despite your objection". Will IBC ever tell us this?
+            // Should we release NFTs / remove the channel from
+            // CHANNELS if this happens?
+            unreachable!("channel can not be closed")
+        }
+        _ => unreachable!("channel can not be closed"),
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_packet_receive(
+    deps: DepsMut,
+    env: Env,
+    msg: IbcPacketReceiveMsg,
+) -> Result<IbcReceiveResponse, Never> {
+    // Regardless of if our processing of this packet works we need to
+    // commit an ACK to the chain. As such, we wrap all handling logic
+    // in a seprate function and on error write out an error ack.
+    match do_ibc_packet_receive(deps, env, msg.packet) {
+        Ok(response) => Ok(response),
+        Err(error) => Ok(IbcReceiveResponse::new()
+            .add_attribute("method", "ibc_packet_receive")
+            .add_attribute("error", error.to_string())
+            .set_ack(ack_fail(&error.to_string()).unwrap())),
+    }
+}
+
+fn do_ibc_packet_receive(
```

Comment:
Given that this function ranges from line 114 to line 311, we will have to refactor this in some degree

Reply:
lol agree. i just love scoping the types like this. :)

Reply:
I would suggest moving the `ibc_packet_receive` entry point into its own file (ibc_receive.rs).

Will be also great for scoping types :)